File: app/src/main/java/com/commit451/gitlab/model/api/RepositoryTreeObject.java
package com.commit451.gitlab.model.api;

import android.support.annotation.StringDef;

import com.squareup.moshi.Json;

import org.parceler.Parcel;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Parcel
public class RepositoryTreeObject {

    public static final String TYPE_FOLDER = "tree";
    public static final String TYPE_REPO = "submodule";
    public static final String TYPE_FILE = "blob";

    @StringDef({TYPE_FOLDER, TYPE_REPO, TYPE_FILE})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Type {
    }

    @Json(name = "id")
    String id;
    @Json(name = "name")
    String name;
    @Json(name = "type")
    @Type
    String type;
    @Json(name = "mode")
    String mode;

    public RepositoryTreeObject() {}

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public @Type String getType() {
        return type;
    }

    public String getMode() {
        return mode;
    }
}

File: app/src/main/java/com/commit451/gitlab/model/api/RepositoryTreeObject.kt
package com.commit451.gitlab.model.api

import android.os.Parcelable
import androidx.annotation.StringDef
import com.squareup.moshi.Json
import kotlinx.android.parcel.Parcelize

@Parcelize
data class RepositoryTreeObject(
    @Json(name = "id")
    var id: String? = null,
    @Json(name = "name")
    var name: String? = null,
    @Json(name = "type")
    @Type
    @get:Type
    var type: String? = null,
    @Json(name = "mode")
    var mode: String? = null
) : Parcelable {
    companion object {

        const val TYPE_FOLDER = "tree"
        const val TYPE_REPO = "submodule"
        const val TYPE_FILE = "blob"
    }

    @StringDef(TYPE_FOLDER, TYPE_REPO, TYPE_FILE)
    @Retention(AnnotationRetention.SOURCE)
    annotation class Type
}
